component/http/cache/cache.go

```diff
@@ -0,0 +1,349 @@
+			}
+		}
+
+		return
+	}
+}
+
+// getResponse will get the appropriate Response either using the cache or the executor,
+// depending on the
+func getResponse(cfg *cacheControl, path, key string, now int64, rc *RouteCache, exec executor) *cachedResponse {
+
+	if cfg.noCache {
+		return exec(now, key)
+	}
+
+	rsp := getFromCache(key, rc)
+	if rsp == nil {
+		metrics.miss(path)
+		response := exec(now, key)
+		return response
+	}
+	if rsp.Err != nil {
+		log.Errorf("error during cache interaction: %v", rsp.Err)
+		metrics.err(path)
+		return exec(now, key)
+	}
+	// if the object has expired
+	if isValid, cx := isValid(now-rsp.LastValid, rc.age.max, append(cfg.validators, cfg.expiryValidator)...); !isValid {
+		tmpRsp := exec(now, key)
+		// if we could not retrieve a fresh Response,
+		// serve the last cached value, with a Warning Header
+		if cfg.forceCache || tmpRsp.Err != nil {
+			rsp.Warning = "last-valid"
+			metrics.hit(path)
+		} else {
+			rsp = tmpRsp
+			metrics.evict(path, cx, now-rsp.LastValid)
+		}
+	} else {
+		// add any Warning generated while parsing the headers
+		rsp.Warning = cfg.warning
+		metrics.hit(path)
+	}
+
+	return rsp
+}
+
+func isValid(age, maxAge int64, validators ...validator) (bool, validationContext) {
+	if len(validators) == 0 {
+		return false, 0
+	}
+	for _, validator := range validators {
+		if isValid, cx := validator(age, maxAge); !isValid {
+			return false, cx
+		}
+	}
+	return true, 0
+}
+
+// getFromCache is the implementation that will provide a cachedResponse instance from the cache,
+// if it exists
+func getFromCache(key string, rc *RouteCache) *cachedResponse {
+	if resp, ok, err := rc.cache.Get(key); ok && err == nil {
+		if b, ok := resp.([]byte); ok {
+			r := &cachedResponse{}
+			err := r.decode(b)
+			if err != nil {
+				return &cachedResponse{Err: fmt.Errorf("could not decode cached bytes as response %v for key %s", resp, key)}
+			}
+			r.FromCache = true
+			return r
+		}
+		// NOTE : we need to do this hack to bypass the redis go client implementation of returning result as string instead of bytes
+		if b, ok := resp.(string); ok {
+			r := &cachedResponse{}
+			err := r.decode([]byte(b))
+			if err != nil {
+				return &cachedResponse{Err: fmt.Errorf("could not decode cached string as response %v for key %s", resp, key)}
+			}
+			r.FromCache = true
+			return r
+		}
+
+		return &cachedResponse{Err: fmt.Errorf("could not parse cached response %v for key %s", resp, key)}
+	} else if err != nil {
+		return &cachedResponse{Err: fmt.Errorf("could not read cache value for [ key = %v , Err = %v ]", key, err)}
+	}
+	return nil
+}
+
+// saveToCache caches the given Response if required with a ttl
+// as we are putting the objects in the cache, if its a TTL one, we need to manage the expiration on our own
+func saveToCache(path, key string, rsp *cachedResponse, cache cache.TTLCache, maxAge time.Duration) {
```

Comment:
since we are in the cache package could we name if just:
```suggestion
func save(path, key string, rsp *cachedResponse, cache cache.TTLCache, maxAge time.Duration) {
```